Added typegen blog (#62)
statelyai/eng-blog: content/posts/2022-01-27-introducing-typegen.mdx

---
title: "Introducing: TypeScript typegen for XState"

> description: "XState's TypeScript experience just got an enormous upgrade - Typescript typege, integrated with VSCode."

Reviewer: ```suggestion
description: "XState's TypeScript experience just got an enormous upgrade - Typescript typegen, integrated with VSCode."
```

> description: "XState's TypeScript experience just got an enormous upgrade - Typescript typege, integrated with VS Code."

Reviewer 2: ```suggestion
description: "XState’s TypeScript experience just got an enormous upgrade - Typescript typegen, integrated with VS Code."
```

tags:
  - announcement
  - xstate
  - community
  - typescript

>   - tutorial

Reviewer: ```suggestion
  - tutorial
  - vscode
```

author:
  - Matt Pocock
originalURL: ""
excerpt: "XState and TypeScript are a match made in heaven, and XState's TypeScript experience just got an enormous upgrade"
publishedAt: "2022-01-27"
---

> **XState and TypeScript are a match made in heaven**. TypeScript gives you type safety, and XState gives you logical safety. Together, they give you confidence that your code will do what you expect it to.

Reviewer 2: ```suggestion
**XState and TypeScript are a match made in heaven**. TypeScript gives you type safety, and XState gives you logical safety. Together, they give you confidence that your code will do what you expect.
```

However, we've been hearing from the community for some time that the _experience_ of using TypeScript with XState needed improving.

Today's your lucky day. **XState's TypeScript experience just got an _enormous_ upgrade**.

> We have brought type generation into XState v4.29.0, and we're aiming for **perfect types on all parts of XState's API**.

Reviewer 2: ```suggestion
We have brought type generation into XState v4.29.0, and we’re aiming for **perfect types on all parts of XState’s API**.
```

> We're releasing it today as an opt-in beta. Be sure to check out the [known limitations](https://xstate.js.org/docs/guides/typescript.html#known-limitations) section of the docs.

Reviewer 2: ```suggestion
We’re releasing typegen today as an opt-in beta. Be sure to check out the [known limitations](https://xstate.js.org/docs/guides/typescript.html#known-limitations) section of the docs.
```

## Getting Started

> You can visit the [official docs](https://xstate.js.org/docs/guides/typescript.html#typegen-with-the-vscode-extension) for the full guide, or follow the instructions here:

Reviewer 2: ```suggestion
You can visit the [official docs](https://xstate.js.org/docs/guides/typescript.html#typegen-with-the-vscode-extension) for the full guide, or follow the instructions:
```

1. Download and install the [VS Code extension](https://marketplace.visualstudio.com/items?itemName=statelyai.stately-vscode).

2. Open a new file and create a new machine, passing the schema attributes:

```ts
import { createMachine } from "xstate";

const machine = createMachine({
  schema: {
    context: {} as { value: string },
    events: {} as { type: "FOO"; value: string } | { type: "BAR" },
  },
  initial: "a",
  states: {
    a: {
      on: {
        FOO: {
          actions: "consoleLogValue",
          target: "b",
        },
      },
    },
    b: {
      entry: "consoleLogValueAgain",
    },
  },
});
```

3. Add `tsTypes: {}` to the machine and save the file:

> ```ts
> const machine = createMachine({
>   tsTypes: {},

Reviewer: ```suggestion
```diff
const machine = createMachine({
+ tsTypes: {},
```

Nit suggestion

  schema: {
    context: {} as { value: string },
    events: {} as { type: "FOO"; value: string } | { type: "BAR" },
  },
  initial: "a",
  states: {
    /* ... */
  },
});
```

4. The extension should automatically add a generic to the machine:

```ts
const machine = createMachine({
  // This generic just got added!
  tsTypes: {} as import("./filename.typegen").Typegen0,
  /* ... */
});
```

> 5. Add a second parameter into the `createMachine` call - this is where you implement the actions, services, guards and delays for the machine.

Reviewer 2: *this* — is the *this* where you implement the actions, services, guards and delays the parameter or the call?

```ts
const machine = createMachine(
  {
    /* ... */
  },
  {
    actions: {
      consoleLogValue: (context, event) => {
        // Wow! event is typed to { type: 'FOO' }
        console.log(event.value);
      },
      consoleLogValueAgain: (context, event) => {
        // Wow! event is typed to { type: 'FOO' }
        console.log(event.value);
      },
    },
  }
);
```

## Typing improvements

> Let's go into the nitty gritty and show you exactly what's improved.

Reviewer 2: ```suggestion
Let’s get into the nitty-gritty and show you exactly what’s improved.
```

### Events in machine options

One of the most common pain points we heard from our community was using named actions, services and guards with TypeScript. The main reason is that you needed to write code like this:

```ts
createMachine(
  {
    schema: {
      events: {} as { type: "FOO"; value: string } | { type: "BAR" },
    },
    on: {
      FOO: {
        actions: "myAction",
      },
    },
  },
  {
    actions: {
      myAction: (context, event) => {
        /**
         * TS don't know if the event is FOO or BAR,
         * so we have to defensively check here
         */
        if (event.type === "FOO") {
          /**
           * Now we know that event.value
           * is present on FOO, because
           * we checked
           */
          console.log(event.value);
        }
      },
    },
  }
);
```

> The VSCode extension **statically analyzes your machine**, and knows which events lead to which actions.

Reviewer: ```suggestion
The VS Code extension **statically analyzes your machine**, and knows which events lead to which actions.
```

```ts
createMachine(
  {
    /* config */
  },
  {
    actions: {
      myAction: (context, event) => {
        /**
         * No more defensive check needed! event
         * is typed to only the events that cause
         * the action
         */
        console.log(event.value);
      },
    },
  }
);
```

This works for actions, services, guards and delays. It even works for **entry actions**, another big pain point from the community.

We're hoping this lets you cut hundreds of lines of useless defensive code.

### Autocomplete on machine options

> Another thing we've been hearing from the community is that it's easy to make typos on machine options.

Reviewer 2: ```suggestion
Another thing we’ve been hearing from the community is that it’s easy to make typos on machine options.
```

> Now, with typegen, you get **autocomplete on all machine options**. This will error:

Reviewer 2: ```suggestion
Now, with typegen, you get **autocomplete on all machine options**. The following code will error:
```

```ts
createMachine(
  {
    entry: ["sayHello"],
  },
  {
    actions: {
      /*
       * This will error, because sayhello does not
       * exist in the machine declaration above
       */
      sayhello: () => {},
    },
  }
);
```

> We've also made it so any _missing_ machine options will error when

Reviewer 2: ```suggestion
We’ve also made it so any _missing_ machine options will error when
```

you implement them later. So, in a React component:

```tsx
const machine = createMachine({
  entry: ["sayHello"],
});

const App = () => {
  /**
   * This will error, because you haven't implemented
   * sayHello in your actions object
   */
  const [state, send] = useMachine(machine);
};
```

> That gives you 100% confidence that your machine has all the things it needs to work.

Reviewer 2: *That* - is *that* the lack of errors?

### Typing of promise-services

Using promise-based services might be the single biggest pain point with XState and TypeScript. We used to have an entire troubleshooting section in our docs dedicated to them.

> Now, you can **strongly type the results of promise-based services**. Here's how:

Reviewer 2: ```suggestion
Now, you can **strongly type the results of promise-based services**. Here’s how:
```

```ts
createMachine(
  {
    schema: {
      /**
       * Pass the 'services' attribute to schema,
       * with all the names of your services and
       * the data they return
       */
      services: {} as {
        myService: {
          // The data that gets returned from the service
          data: { id: string };
        };
      },
    },
    invoke: {
      src: "myService",
      onDone: {
        actions: "consoleLogId",
      },
    },
  },
  {
    services: {
      /**
       * This service will now type error if it
       * returns anything other than { id: string }
       */
      myService: async () => {
        return {
          id: "1",
        };
      },
    },
    actions: {
      consoleLogId: (context, event) => {
        /**
         * This event now knows that it will
         * receive a data attribute with { id: string }
         */
        console.log(event.data.id);
      },
    },
  }
);
```

> This makes handling data return types in XState intuitive, easy and type-safe.

Reviewer 2: *this* - is *this* the strongly-typed results?

## Acknowledgements

> I want to thank my Stately colleague [Andarist](https://twitter.com/AndaristRake). His TypeScript wizardry, incredible attention to detail and deep love of open source helped make this possible. We've literally been talking about this for 18 months - and it's finally here.

Reviewer 2: ```suggestion
I want to thank my Stately colleague [Andarist](https://twitter.com/AndaristRake). His TypeScript wizardry, incredible attention to detail and deep love of open source helped make this possible. We’ve literally been talking about this for 18 months - and it’s finally here.
```
